FXML-638:  LinAlg Linear Operator

cmcgirr-amd opened the pull request:
Extend LinAlg with a new named operator for linear. 
Provide an additional lowering for linear to standard linalg operators. So that the linear operator can be used with testing.

Commit:
feat(LinalgOps.td): add linear operator to linalg

Commit:
feat(Unfuse.cpp): add lowering for the linear operation

allows for the linear op to be compiled with normal linalg

ljfitz reviewed (approved):
It'd be nice not to use `linalg.generic` directly at this moment, but otherwise LGTM.

Commit:
feat(core_named_ops.py): add broadcast_1d_to_2d and transpose2d operations

allows us to explicitly describe what the operator does rather than a convoluted generic op

Commit:
refactor(Unfuse.cpp): use the Transpose2D and Broadcast1DTo2D named ops in the unfuse

cmcgirr-amd commented:
- [ ] Going to let https://github.com/ljfitz/llvm-project/pull/19 merge first and resolve conflicts here

cmcgirr-amd commented on mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp:
```diff
@@ -549,6 +549,78 @@ struct GlobalAveragePool2DLowering : OpRewritePattern<GlobalAveragePool2DOp> {
+    auto biasType = bias.getType().cast<RankedTensorType>();
+    auto outputType = output.getType().cast<RankedTensorType>();
+
+    if (inputType.getRank() != 2 || weightsType.getRank() != 2) {
+      return rewriter.notifyMatchFailure(op,
+                                         "input and weights must be rank 2");
+    }
```
I had these checks included because I noticed the other conversions would check the inputs. And does the verifier run before the pass always? If so then I can remove these checks. 

I also used the rewriter to notify failure because you can get a nice message when using `--debug` to figure out why a pattern did not match rather than an assert because maybe another pattern may match.

ljfitz replied:
`notifyMatchFailure` is great when you don't wish to match some legal IR. I do not yet know when the verifier code is run, but given the fact that it is there I'd prefer if uses only did sanity checking and not match failures for illegal IR.

cmcgirr-amd replied:
Alright, I see your point. I'll change it to asserts for sanity checking. 
But I'm curious what about marking operators/dialect as illegal in the conversion pass to detect unmatched operators?

Commit:
Merge remote-tracking branch 'origin/feature/fused-ops' into christopher.linalg_linear_op.FXML-638

Commit:
refactor(Unfuse.cpp): convert match failures to assert

input bias is handle by the generic broadcast operator so no need to
check. And the ranks of the inputs can simply be an assert. The
verifier should indicate if any operator has inputs that are incorrect.

BraunPhilipp commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -700,6 +700,38 @@ LogicalResult GlobalAveragePool2DOp::verify() {
+  //   * the dim 0 of input must match dim 0 of the output
+  //   * the dim 0 of weights must match dim 1 of the output
+  //   * the dim 0 of bias must match dim 1 of the output
+  if ((inputTensor.getShape()[1] != weightsTensor.getShape()[1]) ||
```
nit: Maybe we can cast to `SmallVector`. Something like `SmallVector<int64_t>(inputTensor.getShape().begin(), inputTensor.getShape().end())` might look slightly nicer and allows a direct equivalence check.
